File: WindFlow/VoipStream/lib/voip_stream/ecr24.cpp
#include "constants.hpp"
#include "ecr24.hpp"
#include "util/log.hpp"

namespace voip_stream {

ECR24::ECR24()
    : filter_(ECR24_NUM_ELEMENTS, ECR24_BUCKETS_PER_ELEMENT, ECR24_BETA, ECR24_BUCKETS_PER_WORD)
{}

void ECR24::operator ()(const Dispatcher::Tuple &tuple, wf::Shipper<FilterTuple> &shipper, wf::RuntimeContext &rc)
{
    DEBUG_LOG("ecr24::tuple " << tuple);

    if (tuple.cdr.call_established) {
        filter_.add(tuple.cdr.calling_number, 1, tuple.cdr.answer_timestamp);
        double ecr = filter_.estimate_count(tuple.cdr.calling_number, tuple.cdr.answer_timestamp);

        FilterTuple result;
        result.ts = tuple.ts;
        result.source = SOURCE_ECR24;
        result.rate = ecr;
        result.cdr = tuple.cdr;
        shipper.push(result);
    }
}

}

File: Storm/VoipStream/src/main/java/VoipStream/ECR24.java
package VoipStream;

import org.apache.storm.task.OutputCollector;
import org.apache.storm.task.TopologyContext;
import org.apache.storm.topology.OutputFieldsDeclarer;
import org.apache.storm.topology.base.BaseRichBolt;
import org.apache.storm.tuple.Fields;
import org.apache.storm.tuple.Tuple;
import org.apache.storm.tuple.Values;
import org.slf4j.Logger;
import common.CallDetailRecord;
import common.Constants;
import common.ODTDBloomFilter;
import common.ScorerMap;
import util.Log;

import java.util.Map;

// TODO basically a copy of ECR with different constants
class ECR24 extends BaseRichBolt {
    private static final Logger LOG = Log.get(ECR24.class);

    private OutputCollector outputCollector;

    private ODTDBloomFilter filter;

    @Override
    public void prepare(Map<String, Object> map, TopologyContext topologyContext, OutputCollector outputCollector) {

        // initialize
        this.outputCollector = outputCollector;

        int numElements = Constants.ECR24_NUM_ELEMENTS;
        int bucketsPerElement = Constants.ECR24_BUCKETS_PER_ELEMENT;
        int bucketsPerWord = Constants.ECR24_BUCKETS_PER_WORD;
        double beta = Constants.ECR24_BETA;
        filter = new ODTDBloomFilter(numElements, bucketsPerElement, beta, bucketsPerWord);
    }

    @Override
    public void declareOutputFields(OutputFieldsDeclarer outputFieldsDeclarer) {
        outputFieldsDeclarer.declare(new Fields("source", "timestamp", "calling_number", "answer_timestamp", "rate", "cdr"));
    }

    @Override
    public void execute(Tuple tuple) {
        LOG.debug("tuple in: {}", tuple);

        long timestamp = (long) tuple.getValueByField("timestamp");
        CallDetailRecord cdr = (CallDetailRecord) tuple.getValueByField("cdr");

        if (cdr.callEstablished) {
            String caller = cdr.callingNumber;
            // add numbers to filters
            filter.add(caller, 1, cdr.answerTimestamp);
            double ecr = filter.estimateCount(caller, cdr.answerTimestamp);

            Values out = new Values(ScorerMap.ECR24, timestamp, caller, cdr.answerTimestamp, ecr, cdr);
            outputCollector.emit(out);
            LOG.debug("tuple out: {}", out);
        }

        //outputCollector.ack(tuple);
    }
}
